test/Test/BaseDAO.hs

```diff
@@ -0,0 +1,271 @@
+  let
+    originateData = OriginateData
+      { odFrom = AddressResolved genesisAddress
+      , odName = "BaseDAO"
+      , odBalance = toMutez 0
+      , odStorage = baseDAOStorage admin bal operators
+      , odContract = baseDAOContract p
+      }
+  dao <- originate originateData
+  pure ((owner1, operator1), (owner2, operator2), dao, admin)
+
+test_BaseDAO :: TestTree
+test_BaseDAO = testGroup "tests to check BaseDAO contract functionality"
+  [ testGroup "Operator:"
+      [ testCase "allows zero transfer from non-existent operator" $
+          nettestTestExpectation zeroTransferScenario
+      , testCase "aborts if there is a failure (due to low balance)" $
+          nettestTestExpectation lowBalanceScenario
+      , testCase "aborts if there is a failure (due to non existent source account)" $
+          nettestTestExpectation noSourceAccountScenario
+      , testCase "aborts if there is a failure (due to bad operator)" $
+          nettestTestExpectation badOperatorScenario
+      , testCase "accepts an empty list of transfers" $
+          nettestTestExpectation emptyTransferListScenario
+      , testCase "validates token id" $
+          nettestTestExpectation validateTokenScenario
+      , testCase "cannot transfer foreign money" $
+          nettestTestExpectation noForeignMoneyScenario
+      ]
+  , testGroup "Owner:"
+      [ testCase "aborts if there is a failure (due to low balance)" $
+          nettestTestExpectation lowBalanceOwnerScenario
+      , testCase "validates token id" $
+          nettestTestExpectation validateTokenOwnerScenario
+      , testCase "cannot transfer foreign money" $
+          nettestTestExpectation noForeignMoneyOwnerScenario
+      , testCase "balanceOf request validates token id" $
+          nettestTestExpectation balanceOfOwnerScenario
+      ]
+  , testGroup "Admin:"
+    [ testCase "transfer tokens from any address to any address" $
+        nettestTestExpectation adminTransferScenario
+    ]
+  ]
+
+zeroTransferScenario :: (Monad m) => NettestImpl m -> m ()
+zeroTransferScenario = uncapsNettest $ do
+  nonexistent :: Address <- newAddress "nonexistent"
+  ((owner1, _), _, dao, _) <- originateBaseDAO SampleFA2.defaultPermissionDescriptor
+  let params =
+        [ ( #from_ .! nonexistent
+          , #txs .! [ (#to_ .! owner1, (#token_id .! 0, #amount .! 0)) ]
+          )
+        ]
+  callFrom (AddressResolved nonexistent) dao (Call @"Transfer") params
+
+lowBalanceScenario :: (Monad m) => NettestImpl m -> m ()
+lowBalanceScenario = uncapsNettest $ do
+  ((owner1, op1), (owner2, _), dao, _) <- originateBaseDAO SampleFA2.defaultPermissionDescriptor
+  let params =
+        [ ( #from_ .! owner1
+          , #txs .! [ (#to_ .! owner2, (#token_id .! 0, #amount .! 200)) ]
+          )
+        ]
+  callFrom (AddressResolved op1) dao (Call @"Transfer") params
+    `expectFailure` (NettestFailedWithError
+      (CustomError #fA2_INSUFFICIENT_BALANCE (#required .! 200, #present .! 100)))
```

Comment:
In latest `cleveland` there are `expectCustomError` and `expectCustomError_` functions, using them should be a bit neater (in ≈all tests).

Reply:
That's definitely neater, I've updated to use these as well.